# Embeddings

- Any piece of text represented as numerical data is **embeddings** or **vectors**.
- Vectors/embeddings are in the form of arrays, i.e., they are known as **vector arrays**.
- **Dimension**: Size (length) of a vector array.
    - E.g., if dimension size = vector array length = 100, then any string, no matter how long or small, output will be 100 length array.

## OpenAI Embedding Model
- OpenAI provides a model to convert text to embeddings. It provides:
    1. **text-embedding-3-small**: Default dimension size of 1,536.
    2. **text-embedding-3-large**: Default dimension size of 3,072.

## Example:
**Sample Text:**
```js
"To make classic milk tea, bring one cup of water to a boil in a small saucepan and add one teaspoon of black tea leaves along with crushed ginger and cardamom. Let it simmer for two to three minutes so the spices infuse thoroughly into the water. Pour in half a cup of whole milk and add sugar to your taste, then bring the mixture back to a rolling boil over medium heat. Once it bubbles up and turns a rich caramel color, remove it from the heat and strain the hot tea through a fine mesh sieve directly into your cup."
```

**Converting to chunks:**
```js
[
    "To make classic milk tea, bring one cup of water to a boil...",
    "Let it simmer for two to three minutes so the spices infuse...",
    "Pour in half a cup of whole milk and add sugar to your taste, ...",
    "Once it bubbles up and turns a rich caramel color, remove it..."
]
```

**Converting to embeddings:**
```js
[
    "To make classic milk tea..." -> [0.0124, -0.0341, ..., 0.0412],
    "Let it simmer for two..."    -> [-0.0051, 0.0218, ..., -0.0073],
    "Pour in half a cup..."       -> [0.0382, -0.0112, ..., 0.0155],
    "Once it bubbles up..."       -> [-0.0189, 0.0045, ..., 0.0319]
]
```

> **Note:** This is an array of arrays (2D array)

> **Also Note:** The size of every chunk is not equal, still all the embeddings are of same size, i.e., 1536 floats

## In Python...

In [35]:
from dotenv import load_dotenv
# Not using OpenAIEmbeddings because: I have no money.
from langchain_google_genai import GoogleGenerativeAIEmbeddings
import numpy as np

load_dotenv()

# Use 'gemini-embedding-001' or 'gemini-embedding-2'
embeddings = GoogleGenerativeAIEmbeddings(model="gemini-embedding-001")

In [27]:
text_chunks = [
    "To make classic milk tea, bring one cup of water to a boil in a small saucepan and add one teaspoon of black tea leaves along with crushed ginger and cardamom.",
    "Let it simmer for two to three minutes so the spices infuse thoroughly into the water.",
    "Pour in half a cup of whole milk and add sugar to your taste, then bring the mixture back to a rolling boil over medium heat.",
    "Once it bubbles up and turns a rich caramel color, remove it from the heat and strain the hot tea through a fine mesh sieve directly into your cup.",
]

In [28]:
vector_array = embeddings.embed_documents(text_chunks)

In [29]:
for vector in vector_array:
    print(f"Dimension: {len(vector)}")
    print(f"    {vector}")

Dimension: 3072
    [0.00024694292, 0.026440745, 0.0022246125, -0.07226342, 0.023469243, 0.0019105234, 0.01983214, 0.004021473, -0.015788851, 0.022819994, -0.0046403497, -0.005303376, 0.0009161539, -0.006404216, 0.114708014, -0.010322176, 0.008528146, -0.001521593, 0.019263232, -0.014517514, -0.014362469, -0.0028901282, 0.009841278, -0.005749436, -0.00060759514, -0.010974117, 0.0108307265, 0.031157162, 0.012749304, -0.006952037, 0.0032184545, 0.023135606, 0.023833705, -0.0032321794, 0.016857754, 0.044616677, 0.020815674, -0.017849263, -0.018309826, 0.010027915, -0.017575787, 0.01777031, -0.022348613, 0.00872617, -0.0024298837, 0.0012523788, 0.016286142, -0.008023805, 0.0033758583, 0.020265859, 0.05041522, 0.02159784, -0.013917123, -0.2131595, -0.012558016, -0.008185853, 0.006039169, -0.037069127, -0.0034397068, 0.0041118017, -0.008612146, -0.01185087, -0.028113531, 0.007872588, -0.028848523, 0.015513415, 0.009282467, 0.023575328, 0.0078104483, -0.016239595, 0.0038262273, 0.027112644, 0

# Cosine Similarity

Cosine similarity is a measure of how much 2 embeddings are similar.

## Working:

### 1. Upload PDF
- Upload a PDF (e.g, Tea Recipe)
- Break into Chunks
- Convert to embeddings
```javascript
[
    "To make classic milk tea..." -> [0.0124, -0.0341, ..., 0.0412],
    "Let it simmer for two..."    -> [-0.0051, 0.0218, ..., -0.0073],
    "Pour in half a cup..."       -> [0.0382, -0.0112, ..., 0.0155],
    "Once it bubbles up..."       -> [-0.0189, 0.0045, ..., 0.0319]
]
```
---
### 2. Query/Prompt
- Enter a query or prompt (e.g., "How much milk required?")
- Convert it to embedding too.
```javascript
    "How much milk required?" -> [0.0354, -0.0108, ..., 0.0162]
```
- Now, the user query and document both are in the same format: Embedding format
---
### 3. Calculate Cosine Similarity
- Compare the user query embedding with **each** PDF chunk embedding to calculate the **cosine similarity score**.
- Cosine Similarity Scores:
    - **Close to 1:** The meanings of the texts are highly similar.
    - **Close to 0:** The meanings of the texts are neutral or unrelated.
    - **Close to -1:** The meanings of the texts are opposite.

<table>
  <thead>
    <tr>
      <th>Chunk Embedding</th>
      <th>Query Embedding</th>
      <th>Cosine Similarity Score</th>
      <th>Verdict</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><code>[0.0124, -0.0341, ..., 0.0412]</code></td>
      <td rowspan="4"><code>[0.0354, -0.0108, ..., 0.0162]</code></td>
      <td>0.42</td>
      <td>Near 0 - Neutral</td>
    </tr>
    <tr>
      <td><code>[-0.0051, 0.0218, ..., -0.0073]</code></td>
      <td>0.31</td>
      <td>Near 0 - Neutral</td>
    </tr>
    <tr>
      <td><code>[0.0382, -0.0112, ..., 0.0155]</code></td>
      <td>0.89</td>
      <td>Near 1 - Similar 🤓</td>
    </tr>
    <tr>
      <td><code>[-0.0189, 0.0045, ..., 0.0319]</code></td>
      <td>0.38</td>
      <td>Near 0 - Neutral</td>
    </tr>
  </tbody>
</table>

---
### 4. Choose top_k
- Choose a **top_k** value, means pick the number of similar embeddings to be chosen
- E.g., $k = 2$, then we pick:
```js
[
    0.89 -> [0.0382, -0.0112, ..., 0.0155] -> "Pour in half a cup..." -> Send to LLM
    0.42 -> [0.0124, -0.0341, ..., 0.0412] -> "To make classic milk tea..." -> Send to LLM
]
```
- LLM receives original text chunks, not the raw embedding numbers.
- This top_k value is not chosen randomly, it undergoes an algorithm known as: **Badme dekhenge**

---
### 5. Send to LLM
- Thus, instead of giving 4 chunks of text to the LLM, we provide only the selected 2, reducing cost.
- This was just a simple example of 4 chunks, in reality
    - There can be 1000s of chunks
    - and only a few of them are picked using cosine simlarity
    - which significantly reduces the cost.
---

## Cosine Similarity Formula
$$\cos(\theta) = \frac{A \cdot B}{\Vert{}A\Vert{} \Vert{}B\Vert{}}$$
Where,
- A: Embedding array of user query
- B: Embedding array of a text chunk

## In Python...

In [79]:
text_chunks = [
    "To make classic milk tea, bring one cup of water to a boil in a small saucepan and add one teaspoon of black tea leaves along with crushed ginger and cardamom.",
    "Let it simmer for two to three minutes so the spices infuse thoroughly into the water.",
    "Pour in half a cup of whole milk and add sugar to your taste, then bring the mixture back to a rolling boil over medium heat.",
    "Once it bubbles up and turns a rich caramel color, remove it from the heat and strain the hot tea through a fine mesh sieve directly into your cup.",
]

user_query = "How much milk and sugar to be added?"

In [ ]:
text_chunk_vector_array = embeddings.embed_documents(text_chunks)

In [80]:
user_query_vector_array = embeddings.embed_query(user_query)

In [81]:
print("---- Text Chunk Vector Array ----")
for vector in text_chunk_vector_array:
    print(f"Dimension: {len(vector)}")
    print(f"    {vector}")

print("\n---- User Query Vector Array ----")
print(f"Dimension: {len(user_query_vector_array)}")
print(f"    {user_query_vector_array}")

---- Text Chunk Vector Array ----
Dimension: 3072
    [0.00024694292, 0.026440745, 0.0022246125, -0.07226342, 0.023469243, 0.0019105234, 0.01983214, 0.004021473, -0.015788851, 0.022819994, -0.0046403497, -0.005303376, 0.0009161539, -0.006404216, 0.114708014, -0.010322176, 0.008528146, -0.001521593, 0.019263232, -0.014517514, -0.014362469, -0.0028901282, 0.009841278, -0.005749436, -0.00060759514, -0.010974117, 0.0108307265, 0.031157162, 0.012749304, -0.006952037, 0.0032184545, 0.023135606, 0.023833705, -0.0032321794, 0.016857754, 0.044616677, 0.020815674, -0.017849263, -0.018309826, 0.010027915, -0.017575787, 0.01777031, -0.022348613, 0.00872617, -0.0024298837, 0.0012523788, 0.016286142, -0.008023805, 0.0033758583, 0.020265859, 0.05041522, 0.02159784, -0.013917123, -0.2131595, -0.012558016, -0.008185853, 0.006039169, -0.037069127, -0.0034397068, 0.0041118017, -0.008612146, -0.01185087, -0.028113531, 0.007872588, -0.028848523, 0.015513415, 0.009282467, 0.023575328, 0.0078104483, -0.01623

### Function for Cosine Similarity
$$\cos(\theta) = \frac{A \cdot B}{\Vert{}A\Vert{} \Vert{}B\Vert{}}$$

In [82]:
def cosine_similarity(a, b):
    a, b = np.array(a), np.array(b)
    return np.dot(a, b) / (np.linalg.norm(a) * np.linalg.norm(b))

### EVERY Text Chunk Embedding v/s User Query Embedding

In [83]:
scores = [
    cosine_similarity(user_query_vector_array, text_chunk_vector_array[0]),
    cosine_similarity(user_query_vector_array, text_chunk_vector_array[1]),
    cosine_similarity(user_query_vector_array, text_chunk_vector_array[2]),
    cosine_similarity(user_query_vector_array, text_chunk_vector_array[3]),
]

print(scores)

[np.float64(0.7331218705864361), np.float64(0.6830856987976053), np.float64(0.7651386612683204), np.float64(0.6612680525628856)]


In [84]:
max_score = max(scores)
print(f"The max score is: {max_score}")

max_score_index = scores.index(max_score)
print(f"The index of the max score is: {max_score_index}")

The max score is: 0.7651386612683204
The index of the max score is: 2


In [85]:
print(f"Thus, the most matching chunk is:\n\n'{text_chunks[max_score_index]}'")

Thus, the most matching chunk is:

'Pour in half a cup of whole milk and add sugar to your taste, then bring the mixture back to a rolling boil over medium heat.'


... Thus, we will provide only this chunk to LLM instead of giving all the 4 chunks.

### If we want to find top K similarity scores, not just 1...

In [86]:
k = 2
top_k_indices = np.argsort(scores)[::-1][:k]
# [::-1] sorts the array in descending order (large to small)
# [:k] picks the first k

top_k_indices

array([2, 0])

In [87]:
print("--- Matching Chunks ---")
matching_text_chunks = []

for index in top_k_indices:
    matching_text_chunks.append(text_chunks[index])

matching_text_chunks

--- Matching Chunks ---


['Pour in half a cup of whole milk and add sugar to your taste, then bring the mixture back to a rolling boil over medium heat.',
 'To make classic milk tea, bring one cup of water to a boil in a small saucepan and add one teaspoon of black tea leaves along with crushed ginger and cardamom.']

# Conclusion:

- Finding the top matching text chunks = **Retrieval**
- Combining retrieved text chunks + user query and giving to LLM = **Augmentation**
- LLM generating output = **Generation**

Thus, RAG is complete (see below)

In [88]:
from langchain_google_genai import ChatGoogleGenerativeAI
from langchain_core.messages import SystemMessage, HumanMessage

llm = ChatGoogleGenerativeAI(model="gemini-2.5-flash")

messages = [
    SystemMessage(
        """
            Answer only with the given context.
            Do not make up answer.
            If question is out of context, reply "I don't know"
        """
    ),
    HumanMessage(
        f"""
        Context: {matching_text_chunks}
        Que: {user_query}
        """
    )
]

result = llm.invoke(messages)

result.content

'Pour in half a cup of whole milk and add sugar to your taste.'